File: algorithms/dynamic-programming/sequence-alignment/c/sequencealignment.c
#include <stdio.h>
#include <string.h>

#define GAP_COST 4
#define MISMATCH_COST 3

int min(int a, int b, int c) {
    int m = a;
    if (b < m) m = b;
    if (c < m) m = c;
    return m;
}

int sequence_alignment(const char *s1, const char *s2) {
    int m = strlen(s1);
    int n = strlen(s2);

    int dp[m + 1][n + 1];

    for (int i = 0; i <= m; i++)
        dp[i][0] = i * GAP_COST;
    for (int j = 0; j <= n; j++)
        dp[0][j] = j * GAP_COST;

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            int match_cost = (s1[i - 1] == s2[j - 1]) ? 0 : MISMATCH_COST;
            dp[i][j] = min(
                dp[i - 1][j - 1] + match_cost,
                dp[i - 1][j] + GAP_COST,
                dp[i][j - 1] + GAP_COST
            );
        }
    }

    return dp[m][n];
}

int main() {
    printf("%d\n", sequence_alignment("GCCCTAGCG", "GCGCAATG")); // 18
    return 0;
}

File: algorithms/dynamic-programming/sequence-alignment/rust/sequence_alignment.rs
use std::cmp;

const GAP_COST: i32 = 4;
const MISMATCH_COST: i32 = 3;

pub fn sequence_alignment(s1: &str, s2: &str) -> i32 {
    let m = s1.len();
    let n = s2.len();
    let s1_bytes = s1.as_bytes();
    let s2_bytes = s2.as_bytes();

    let mut dp = vec![vec![0i32; n + 1]; m + 1];

    for i in 0..=m {
        dp[i][0] = i as i32 * GAP_COST;
    }
    for j in 0..=n {
        dp[0][j] = j as i32 * GAP_COST;
    }

    for i in 1..=m {
        for j in 1..=n {
            let match_cost = if s1_bytes[i - 1] == s2_bytes[j - 1] { 0 } else { MISMATCH_COST };
            dp[i][j] = cmp::min(
                cmp::min(dp[i - 1][j] + GAP_COST, dp[i][j - 1] + GAP_COST),
                dp[i - 1][j - 1] + match_cost,
            );
        }
    }

    dp[m][n]
}

fn main() {
    println!("{}", sequence_alignment("GCCCTAGCG", "GCGCAATG")); // 18
}
